drivers/sensor/fxos8700: Fix racing conditions during initialization

diegosueiro opened the pull request:
Two racing conditions were identified during the initialization of the
drver when the trigger function was enabled:

1 - The fxos8700_handle_int was trying to acquire the semaphore before
    it gets initialized. To solve this we need to initialize the
    semaphore before calling the fxos8700_trigger_init function.

2 - During the fxos8700_trigger initialization the i2c bus is used at
    the same time as the fxos8700_set_power is being called. To fix
    this we need to acquire the semaphore before calling
    fxos8700_set_power function.

These two scenarios was reproducible in the WaRP7 board with i.MX7 SoC.

Signed-off-by: Diego Sueiro <diego.sueiro@gmail.com>

MaureenHelm commented on drivers/sensor/fxos8700/fxos8700.c:
```diff
@@ -332,20 +332,21 @@ static int fxos8700_init(struct device *dev)
 		return -EIO;
 	}
 
+	k_sem_init(&data->sem, 1, UINT_MAX);
```
You can instead initialize the count to zero and remove the `k_sem_take` in line 343.

MaureenHelm reviewed (changes requested):
I don't quite understand the second point in your commit message.  `fxos8700_trigger_init` and `fxos8700_set_power` are called in series from the same thread, so how are they using the i2c bus at the same time?

diegosueiro commented:
@MaureenHelm,

> I don't quite understand the second point in your commit message. fxos8700_trigger_init and fxos8700_set_power are called in series from the same thread, so how are they using the i2c bus at the same time?

I could reproduce the racing condition in the `fxos8700_handle_int` not in the `fxos8700_trigger_init`. The function calling chain was `fxos8700_trigger_init->fxos8700_thread_main->fxos8700_handle_int->i2c_reg_read_byte` with the last function being interrupted by `fxos8700_set_power` from `fxos8700_init`.

diegosueiro commented:
@MaureenHelm ,

I attached two screenshots to demonstrate this race condition.
In the channel 0 I have a GPIO going up/down around the ``i2c_reg_read_byte`` in ``fxos8700_handle_int`` function and in the channel 1 another GPIO going up/down around the ``fxos8700_set_power`` in ``fxos8700_int`` function.

The first screenshot is for calling ``fxos8700_set_power`` without the semaphore protection and the second is with the semaphore protection.


![fxos8700_wo_sem_protect](https://user-images.githubusercontent.com/598588/43880538-c7881510-9ba0-11e8-9ed9-24c98eb7e52d.png)


![fxos8700_sem_protect](https://user-images.githubusercontent.com/598588/43880537-c76c67d4-9ba0-11e8-82e7-ef162e6c826d.png)


As you cann see from the first screenshot the i2c functions never return.

MaureenHelm commented:
I see only one race condition: `fxos8700_handle_int` runs before `fxos8700_init` is complete. This reveals two problems:
1) The semaphore hasn't been initialized in time for `fxos8700_handle_int`. This can be fixed by initializing the semaphore earlier, as you've already done.
2) `fxos8700_init` assumes it has control of the i2c bus when it invokes `fxos8700_set_power`. This can be fixed by initializing the semaphore count to 0 and giving the semaphore at the end of the function. That will allow the sensor initialization to finish before processing any triggers.

diegosueiro committed:
drivers/sensor/fxos8700: Fix race condition during initialization

A race condition was identified during the initialization of the
driver when the trigger function is enabled. This exposed two issues:

1 - The fxos8700_handle_int is trying to acquire the semaphore before
    it gets initialized. To solve this we need to initialize the
    semaphore before calling the fxos8700_trigger_init function.

2 - During the fxos8700_trigger initialization the i2c bus is used
    (from fxos8700_handle_int function) at the same time as the
    fxos8700_set_power is being called. To fix this we need to
    initialize the semaphore with 0 and just release it after calling
    the fxos8700_set_power function in the fxos8700_int

These two scenarios was reproducible in the WaRP7 board with i.MX7 SoC.

Signed-off-by: Diego Sueiro <diego.sueiro@gmail.com>

MaureenHelm commented:
recheck
